content/docs/tutorials/connect-php-app.mdx

```diff
@@ -0,0 +1,172 @@
+---
+title: 'Connect a PHP application to PlanetScale'
+subtitle: 'Spin up a PlanetScale MySQL serverless database in seconds and connect to a PHP application'
+---
+
+## Introduction
+
+In this tutorial, you'll learn how to connect a PHP application to a PlanetScale MySQL database using the PDO abstraction class.
+
+## Prerequisites
+
+- [PHP](https://www.php.net/manual/en/install.php) &mdash; This tutorial uses `v8.0`
+- A [free PlanetScale account](https://auth.planetscale.com/sign-up)
+
+## Set up the PHP app
+
+1. Create your index.php file (or any php file is fine).
+
+```
+<?php 
+
+//Your code will go here
+
+?>
+```
+
+## Set up the database
+
+Next, you need to set up your PlanetScale database and connect to it.
+
+You can create a database either in the [PlanetScale dashboard](https://app.planetscale.com) or from the PlanetScale CLI. This guide will use the dashboard, you can follow the database setup instructions in the [PlanetScale quick start guide](/tutorials/planetscale-quick-start-guide#quick-start-with-the-planetscale-application).
+
+1. Create a new database with a default 'main' branch
+
+That's it! Your database is ready to use. Next, let's connect it to the PHP application and then add some data.
+
+## Connect to the PHP app
+
+There are two ways to connect your PHP app to PlanetScale.
+
+-Using the mysqli class
+-Using the PDO class
+
+Both options are covered below.
+
+### Option 1: Using mysqli class
+
+```
+$mysqli = mysqli_init();
+$mysqli->ssl_set(NULL, NULL, "/etc/ssl/cert.pem", NULL, NULL);
+$mysqli->real_connect($_ENV["HOST"], $_ENV["USERNAME"], $_ENV["PASSWORD"], $_ENV["DATABASE"]);
+$mysqli->close();
+```
+
+this is assuming you are using an .env file to store your database information
+
+example ENV file:
+```
+DB_CONNECTION=mysql
+DB_HOST=<ACCESS HOST URL> // this can be found in your planetscale dashboard
+DB_PORT=3306
+DB_DATABASE=<DATABASE_NAME> // what you named the database
+DB_USERNAME=<USERNAME> // this can be found in your planetscale dashboard
+DB_PASSWORD=<PLAIN TEXT> // this can be found in your planetscale dashboard
+MYSQL_ATTR_SSL_CA=/etc/ssl/cert.pem
+```
+
+### Option 2: Using the PDO class
+I tend to prefer using PDO objects as a personal preference.
+Using PDO with SSL requires us to also setup additional 'options' when instantiating the handle.
+
+```
+$dsn = 'mysql:host=$DB_HOST;dbname=$$DB_DATABASE;port=3306';
+$user = "";
+$dbP = "";
+$options = array(
+    1009 => "/etc/ssl/cert.pem",
+    1014 => false,
+);
+try {
+    $db = new PDO($dsn, $user, $dbP, $options);
+    $db->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
+    echo "Connected successfully";
+} catch (PDOException $error) {
+    $msg = $error->getMessage();
+    echo "error:".$msg;
+}
+```
+
+The above snippet is a more concise way of setting these options:
+
+```
+$options = array(
+    PDO::MYSQL_ATTR_SSL_CA => '/etc/ssl/certs/ca-certificates.crt',
+    PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT => false,
```

Comment:
Hi @IAmLuisJ. First of all thanks for adding this and your article on how to connect to PlanetScale. I can't really speak for whether our documentation wants to get this merged or not, but I can speak on this specific attribute.

I also mentioned on Twitter in https://twitter.com/dbussink/status/1507594562407124992, but this option **must** be set to `true`. 

Setting this to `false` means that any connection is insecure and susceptible to a [Man-in-the-middle-attack](https://twitter.com/dbussink/status/1507594562407124992) because the certificate is not verified. 

You mentioned in your article at https://dev.to/iamluisj/connecting-to-planetscale-with-a-pdo-object-in-php-2h09 that you hit a problem when you didn't set that option. I'd like to know more about why that was happening, because I don't understand how setting this option would resolve the issue you mentioned there. 

The error `openssl s_client no cipher match` would indicate there's no match in the TLS cipher that is negotiated during the connection. The cipher negotiation phase happens before the certificate validation step and is unrelated to that step. So changing the verification option in normal circumstances could not impact the cipher selection mechanism which is why I don't understand how this helped with any issue you hit.

Normally that error could happen on really old versions of OpenSSL that don't support modern ciphers. Do you have the exact details of PHP version, OpenSSL version etc. that you tried when you hit that problem? You also mentioned "you may" so it's not consistent? You only had it on specific versions then? 

I'd love to get to the bottom of this, also to get your article updated then. We care deeply about the security of our users and really want to work with people who create community content to make sure they also recommend a secure way to connect to PlanetScale. So anything we can do to help you to get your article updated we're happy to help with. We wouldn't want people to end up on your page as the answer to "how to connect with PDO to PlanetScale" and then not have a secure setup.

Reply:
Btw, we also have an internal (to be hopefully open sourced in the not too far future) repository with examples that we use which does include a PDO example. It looks like this:

```php
<?php
error_reporting(E_STRICT);

$dsn = "mysql:host=[HOSTNAME];dbname=[DATABASE]";
$options = array(
  PDO::MYSQL_ATTR_SSL_CA => "/etc/ssl/certs/ca-certificates.crt",
);

$pdo = new PDO($dsn, "[USERNAME]", "[PASSWORD]", $options);
?>
Successfully connected to PlanetScale!
```

It doesn't set the `PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT` attribute since it then defaults to `true`. It also doesn't use the magic constant values since those are often hard to understand / know what they mean later on so it uses the explicitly named constant also for the certificate path.

Reply:
I am using PHP version 8.1
Omitting the PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT option gives error: error:SQLSTATE[HY000] [1105] unknown error: Code: UNAVAILABLE server does not allow insecure connections, client must use SSL/TLS

Setting it to true gives error:
error:SQLSTATE[HY000] [2026] SSL connection error: no cipher match

OpenSSL info:
OpenSSL support | enabled
-- | --
OpenSSL Library Version | OpenSSL 1.1.1d 10 Sep 2019
OpenSSL Header Version | OpenSSL 1.1.1d 10 Sep 2019
Openssl default config | /opt/alt/openssl11/etc/pki/tls/openssl.cnf

Reply:
@IAmLuisJ What operating system (and distribution for Linux) are you using? And which version(s)? I'd like to be able to reproduce the issue then. 

> error:SQLSTATE[HY000] [2026] SSL connection error: no cipher match

This one still baffles me since I don't know how this can happen :confused:.

Reply:
System | Linux premium158.web-hosting.com 2.6.32-954.3.5.lve1.4.79.1.el6.x86_64 #1 SMP Tue Jul 28 05:24:18 EDT 2020 x86_64
-- | --
Build Date | Mar 15 2022 20:55:22
Build System | Linux buildfarm02.corp.cloudlinux.com 4.18.0-348.7.1.el8_5.x86_64 #1 SMP Tue Dec 21 13:57:48 EST 2021 x86_64 x86_64 x86_64 GNU/Linux

Is there a specific cipher suite I should try setting with the PDO::MYSQL_ATTR_SSL_CIPHER attribute?

I found [this other article](https://fuelphp.com/forums/discussion/12659/using-ssl-with-pdo-mysql-connections/p1) that had the same issue.
"Important is the last constant, in case of self signed SSL certificates (the default when using mysql_ssl_rsa_setup)."

Reply:
> Is there a specific cipher suite I should try setting with the PDO::MYSQL_ATTR_SSL_CIPHER attribute?

No, the defaults should be fine and this doesn't need to be set.

> System Linux premium158.web-hosting.com 2.6.32-954.3.5.lve1.4.79.1.el6.x86_64 https://github.com/planetscale/docs/pull/1 SMP Tue Jul 28 05:24:18 EDT 2020 x86_64

This looks like RedHat Enterprise Linux, but it does look like a really old version. Based on the version number it looks like RedHat Enterprise Linux 6, but the custom paths you mentioned earlier seem like a custom installed PHP. Is this a server of yourself or a shared hosted solution you're using?

RedHat also means that you're probably using the wrong setting for `PDO::MYSQL_ATTR_SSL_CA`. Does the `/etc/ssl/certs/ca-certificates.crt` path actually exist? 

Our documentation at https://docs.planetscale.com/concepts/secure-connections calls out the following path for RedHat:

```
/etc/pki/tls/certs/ca-bundle.crt
```

Can you try then with the following options:

```
$options = array(
  PDO::MYSQL_ATTR_SSL_CA => "/etc/pki/tls/certs/ca-bundle.crt",
  PDO:: MYSQL_ATTR_SSL_VERIFY_SERVER_CERT => true,
);
```

Reply:
This is a shared VPS hosted on namecheap.

With those options I get:
error:SQLSTATE[HY000] [2026] SSL connection error: no cipher match

Same error trying to use the string out of 
var_dump(openssl_get_cert_locations());
default_cert_file: (string) "/opt/alt/openssl11/etc/pki/tls/cert.pem"
default_cert_file_env: (string) "SSL_CERT_FILE"
default_cert_dir: (string) "/opt/alt/openssl11/etc/pki/tls/certs"
default_cert_dir_env: (string) "SSL_CERT_DIR"
default_private_dir: (string) "/opt/alt/openssl11/etc/pki/tls/private"
default_default_cert_area: (string) "/opt/alt/openssl11/etc/pki/tls"
ini_cafile: (string) ""
ini_capath: (string) ""

Reply:
@IAmLuisJ Ah ok, did you do any manual setup yourself? I kinda would want to see if there's a way for me to get access to something with the exact same setup to find out what's going on, since I'm unable to reproduce this problem and it works all fine for me (which is also why our internal earlier example looks the way it looks). 

> default_cert_file: (string) "/opt/alt/openssl11/etc/pki/tls/cert.pem"

I guess the same error if you use this path? And this path does exist with a whole number of root CAs?

Reply:
No manual setup really, purchased hosting from namecheap, changed PHP version to 8.1 through Cpanel (I think it defaulted to 7.4).
I have another shared server through GoDaddy and I was able to get that working by only setting
$options = array(
    1009 => "/opt/alt/openssl11/etc/pki/tls/cert.pem"
);

GoDaddy server info:


PHP Version 8.0.13
--
System | Linux p3plzcpnl467530.prod.phx3.secureserver.net 2.6.32-954.3.5.lve1.4.86.el6.x86_64 #1 SMP Tue Aug 31 17:08:39 UTC 2021 x86_64
-- | --
Build Date | Nov 16 2021 22:47:51
Build System | Linux buildfarm03 4.18.0-305.19.1.el8_4.x86_64 #1 SMP Wed Sep 15 15:39:39 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux



Just for testing I tried setting PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT to false and got the following error:
error:SQLSTATE[HY000] [2026] SSL connection error: Failed to set ciphers to use

So this may just be something weird with my server.  I've updated my [article]() to include a warning about man-in-the-middle-attacks:

Reply:
@IAmLuisJ This is all so very strange. Just tried setting `PDO::MYSQL_ATTR_SSL_VERIFY_SERVER_CERT => true` on my Ubuntu 20.04 machine and it all works perfectly and as expected. 

This must be something about how the shared server is set up and something more specific about that environment that triggers these errors. I have no clue what it would be though :confused:.

Reply:
@IAmLuisJ Our user interface now also shows how to connect with PDO in PHP btw, next to how to use `mysqli`.